collectors/cgroups.plugin/sys_fs_cgroup.c

```diff
@@ -73,6 +73,129 @@ static uint32_t Write_hash = 0;
 static uint32_t user_hash = 0;
 static uint32_t system_hash = 0;
 
+enum cgroups_type { CGROUPS_AUTODETECT_FAIL, CGROUPS_V1, CGROUPS_V2 };
+
+enum cgroups_systemd_setting {
+    SYSTEMD_CGROUP_ERR,
+    SYSTEMD_CGROUP_LEGACY,
+    SYSTEMD_CGROUP_HYBRID,
+    SYSTEMD_CGROUP_UNIFIED
+};
+
+struct cgroups_systemd_config_setting {
+    char *name;
+    enum cgroups_systemd_setting setting;
+};
+
+static struct cgroups_systemd_config_setting cgroups_systemd_options[] = {
+    { .name = "legacy",  .setting = SYSTEMD_CGROUP_LEGACY  },
+    { .name = "hybrid",  .setting = SYSTEMD_CGROUP_HYBRID  },
+    { .name = "unified", .setting = SYSTEMD_CGROUP_UNIFIED },
+    { .name = NULL,      .setting = SYSTEMD_CGROUP_ERR     },
+};
+
+/* on Fed systemd is not in PATH for some reason */
+#define SYSTEMD_CMD_RHEL "/usr/lib/systemd/systemd --version"
+#define SYSTEMD_HIERARCHY_STRING "default-hierarchy="
+
+#define MAXSIZE_PROC_CMDLINE 4096
+static enum cgroups_systemd_setting cgroups_detect_systemd(const char *exec)
+{
+    pid_t command_pid;
+    enum cgroups_systemd_setting retval = SYSTEMD_CGROUP_ERR;
+    char buf[MAXSIZE_PROC_CMDLINE];
+    char *begin, *end;
+
+    FILE *f = mypopen(exec, &command_pid);
+
+    if (!f)
+        return retval;
+
+    while (fgets(buf, MAXSIZE_PROC_CMDLINE, f) != NULL) {
+        if ((begin = strstr(buf, SYSTEMD_HIERARCHY_STRING))) {
+            end = begin = begin + strlen(SYSTEMD_HIERARCHY_STRING);
+            if (!*begin)
+                break;
+            while (isalpha(*end))
+                end++;
+            *end = 0;
+            for (int i = 0; cgroups_systemd_options[i].name; i++) {
+                if (!strcmp(begin, cgroups_systemd_options[i].name)) {
+                    retval = cgroups_systemd_options[i].setting;
+                    break;
+                }
+            }
+            break;
+        }
+    }
+
+    if (mypclose(f, command_pid))
+        return SYSTEMD_CGROUP_ERR;
+
+    return retval;
+}
+
+static enum cgroups_type cgroups_try_detect_version()
+{
+    pid_t command_pid;
+    char buf[MAXSIZE_PROC_CMDLINE];
+    enum cgroups_systemd_setting systemd_setting;
+    int cgroups2_available = 0;
+
+    // 1. check if cgroups2 availible on system at all
+    FILE *f = mypopen("grep cgroup /proc/filesystems", &command_pid);
+    if (!f) {
+        error("popen failed");
+        return CGROUPS_AUTODETECT_FAIL;
+    }
+    while (fgets(buf, MAXSIZE_PROC_CMDLINE, f) != NULL) {
+        if (strstr(buf, "cgroup2")) {
+            cgroups2_available = 1;
+            break;
+        }
+    }
+    if(mypclose(f, command_pid))
+        return CGROUPS_AUTODETECT_FAIL;
+
+    if(!cgroups2_available)
+        return CGROUPS_V1;
+
+    // 2. check systemd compiletime setting
```

Comment:
Netdata runs on systems that does not use `systemd`, is really necessary to try to use the next calls on them?

Reply:
well in case system has no `systemd` then autodetection fails and we return to old behavior (default legacy cgroups).
Not sure what advantage will give us If we prevent trying to exec `systemd`.